kytos/core/link.py

```diff
@@ -121,65 +123,55 @@ def id(self):  # pylint: disable=invalid-name
         """
         return self._id
 
-    @property
-    def available_tags(self):
+    def available_tags(self, tag_type: str = 'vlan') -> list[list[int]]:
         """Return the available tags for the link.
 
         Based on the endpoint tags.
         """
-        return [tag for tag in self.endpoint_a.available_tags if tag in
-                self.endpoint_b.available_tags]
-
-    def use_tag(self, tag):
-        """Remove a specific tag from available_tags if it is there.
-
-        Deprecated: use only the get_next_available_tag method.
-        """
-        if self.is_tag_available(tag):
-            self.endpoint_a.use_tag(tag)
-            self.endpoint_b.use_tag(tag)
-            return True
-        return False
+        return Interface.range_intersection(
+            self.endpoint_a.available_tags[tag_type],
+            self.endpoint_b.available_tags[tag_type],
+        )
 
-    def is_tag_available(self, tag):
+    def is_tag_available(self, tag: int, tag_type: str = 'vlan'):
         """Check if a tag is available."""
-        return (self.endpoint_a.is_tag_available(tag) and
-                self.endpoint_b.is_tag_available(tag))
-
-    def get_next_available_tag(self):
+        return (self.endpoint_a.is_tag_available(tag, tag_type) and
+                self.endpoint_b.is_tag_available(tag, tag_type))
+
+    def get_next_available_tag(
+        self,
+        controller,
+        tag_type: str = 'vlan'
+    ) -> int:
         """Return the next available tag if exists."""
         with self._get_available_vlans_lock:
```

Comment:
`_get_available_vlans_lock[link.id]`, ideally should be a `defaultdict(Lock)` just so it can lock per link optimally. If you could take the opportunity to enhance this part, that'd be great.

Reply:
Changed to `defaultdict(Lock)`